sprockets/mixins/amqp/amqp.py

```diff
@@ -0,0 +1,448 @@
+                         'the timeout (%i seconds)', self._timeout)
+            if self._on_persistent_failure:
+                self._on_persistent_failure(
+                    self,
+                    exchange,
+                    routing_key,
+                    body,
+                    properties
+                )
+            else:
+                raise web.HTTPError(504, 'AMQP connection timeout')
+
+    @property
+    def blocked(self):
+        """Returns ``True`` if the connection is blocked by RabbitMQ.
+
+        :rtype: bool
+
+        """
+        return self._state == self.STATE_BLOCKED
+
+    @property
+    def closed(self):
+        """Returns ``True`` if the connection to RabbitMQ is closed.
+
+        :rtype: bool
+
+        """
+        return self._state == self.STATE_CLOSED
+
+    @property
+    def closing(self):
+        """Returns ``True`` if the connection to RabbitMQ is closing.
+
+        :rtype: bool
+
+        """
+        return self._state == self.STATE_CLOSING
+
+    @property
+    def ready(self):
+        """Returns ``True`` if the connection to RabbitMQ is established and
+        we can publish to it.
+
+        :rtype: bool
+
+        """
+        return self._state in (self.STATE_CONNECTED, self.STATE_BLOCKED)
+
+    @property
+    def connected(self):
+        """Returns ``True`` if the connection to RabbitMQ is connected.
+
+        :rtype: bool
+
+        """
+        return self._state == self.STATE_CONNECTED
+
+    @property
+    def channel(self):
+        """Return the currently opened channel
+
+        :rtype pika.channel.Channel: The channel open with RabbitMQ
+
+        """
+        return self._channel
+
+    @property
+    def _parameters(self):
+        """Return a pika URLParameters object using the AMQP url
+        which identifies the Rabbit MQ server as a URL ready for
+        :class:`pika.connection.URLParameters`.
+
+        :rtype: pika.URLParameters
+
+        """
+        parameters = pika.URLParameters(self._url)
+        if self._connection_attempts > 0:
+            parameters.connection_attempts = self._connection_attempts
+        return parameters
+
+    def connect(self):
+        """This method connects to RabbitMQ, returning the connection handle.
+        When the connection is established, the on_connection_open method
+        will be invoked by pika.
+
+        :rtype: pika.TornadoConnection
+
+        """
+        LOGGER.debug('Connecting to %s', self._url)
+        return pika.TornadoConnection(
+            parameters=self._parameters,
+            on_open_callback=self._on_connection_open,
+            on_open_error_callback=self._on_connection_open_error,
+            on_close_callback=self._on_connection_closed,
+            custom_ioloop=self._ioloop)
+
+    def close(self):
+        """Cleanly shutdown the connection to RabbitMQ
+
+        """
+        LOGGER.debug('Stopping')
+        if not self.ready:
```

Comment:
I don't think this will cleanly close in all circumstances. If close is called in the middle of a connection / reconnection attempt, the call to close will fail and the reconnection will occur. Probably just an edge case, but worth noting.